[RFC] Add os_copy function that uses uv_fs_copyfile (Issue #8288)

Said6289 opened the pull request:
closes #8288

Said6289 committed:
fs: add os_copy function that uses uv_fs_copyfile

Said6289 committed:
fs: replace another custom copy with os_copy

Said6289 committed:
fs: remove unecessary copybuf and os_open call

justinmk commented on src/nvim/fileio.c:
```diff
@@ -2792,8 +2784,8 @@ buf_write (
          */
-        (void)copy_option_part(&dirp, copybuf, BUFSIZE, ",");
-        rootname = get_file_in_dir(fname, copybuf);
+        (void)copy_option_part(&dirp, IObuff, IOSIZE, ",");
```
Curious about why you removed `copybuf`. Is it because the code removed below means `IObuff` is not used for other steps? Might be less risky to keep `copybuf`, otherwise we need to make sure `IObuff` is not used by any other codepaths in this area.

Said6289 replied:
As far as I understand `copybuf` was being used for the copy operation which is now taken care of by libuv. That's why I removed it. It was, however, being used as an argument to `copy_option_part`, but then I found `copy_option_part` being called this way at https://github.com/Said6289/neovim/blob/master/src/nvim/fileio.c#L2911

justinmk commented:
Nice! This looks like a nice code reduction. One small(?) regression is that if it's a big file, the user can't CTRL-c to cancel the operation after this, because we removed `os_breakcheck()` from the copy routine. On systems which support "instant" copy that of course isn't a problem. 

The code reduction is worth the potential regression, I think.

Should we also set `UV_FS_COPYFILE_FICLONE` flag? ref: https://github.com/libuv/libuv/pull/1491

Said6289 committed:
fs: add UV_FS_COPYFILE_FICLONE flag to os_copy

Said6289 committed:
style: make linter happy with fileio.c

Said6289 committed:
test: make first attempt at some kind of test

Said6289 commented:
I made an attempt at writing a test, but for some reason, another test case is now crashing. Am I going in the right direction? Is there a way I can locally check code coverage?

justinmk commented:
> attempt at writing a test, but for some reason, another test case is now crashing

Where? Travis CI passed. Only AppVeyor failed, and it's just a line-ending issue:

```
  [  FAILED  ] 1 test, listed below:
  [  FAILED  ] C:/projects/neovim/test/functional\core\fileio_spec.lua @ 70: fileio creates a backup
  C:/projects/neovim/test/functional\core\fileio_spec.lua:85: Expected objects to be the same.
  Passed in:
  (string) 'foobar
  
  '
  Expected:
  (string) 'foobar
  '
```

justinmk commented on test/functional/core/fileio_spec.lua:
```diff
@@ -64,5 +66,25 @@ describe('fileio', function()
+    local bar_contents = read_file('Xtest_startup_file1~')
+
+    eq('foobar\n', foo_contents);
+    eq('foo\n', bar_contents);
```
on Windows lines end with `\r\n`. To make this test cross-platform, could just trim whitespace. 

There is a trim() function in `test/unit/helpers.lua`: 

https://github.com/neovim/neovim/blob/230a4fc2c327a24bd9515304c9197f8fb91313c7/test/unit/helpers.lua#L123                                                   

We need to lift that up into `test/helpers.lua` so that functional tests can use it.

Said6289 replied:
I forgot about that. I moved the trim function to `test/helpers.lua` and used it in my test.

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -643,6 +643,20 @@ ptrdiff_t os_write(const int fd, const char *const buf, const size_t size,
+/// @param path Path of file to be copied
+/// @param path_new Path of new file
+/// @param flags Bitwise OR of flags defined in <uv.h>
+/// @return libuv error code on error
```
/// @return 0 on success, or libuv error code on failure.

Said6289 committed:
test: move trim to global helpers

Said6289 committed:
docs: explicitly state return value on success

Said6289 commented:
> Where? Travis CI passed. Only AppVeyor failed, and it's just a line-ending issue:

This test is now failing on my machine:
```
[  FAILED  ] test/functional/eval/executable_spec.lua @ 56: executable() exec-bit set, unqualified and not in $PATH
test/functional/eval/executable_spec.lua:58: Expected objects to be the same.
Passed in:
(number) 1
Expected:
(number) 0
```

I'm not sure how my changes could have caused that. Maybe this is an unrelated bug?

justinmk commented on src/nvim/fileio.c:
```diff
@@ -2855,87 +2847,30 @@ buf_write (
-            if (file_info_new.stat.st_gid != file_info_old.stat.st_gid
-                && os_fchown(bfd, -1, file_info_old.stat.st_gid) != 0) {
-              os_setperm((const char *)backup,
-                         (perm & 0707) | ((perm & 07) << 3));
```
I think we still need this step. see https://github.com/libuv/libuv/pull/1547 
libuv copies the perm bits but not the group name.

justinmk committed:
Merge #9709 'fileio: use os_copy to create backups'

ref #8288

justinmk committed:
fileio: set group of backup file

Restores code removed in #9709.

uv_fs_copyfile() copies the perm bits but not the group name.
https://github.com/libuv/libuv/pull/1547

ref #9709
ref #8288

justinmk committed:
lua/shared: share trim() impl

justinmk commented on src/nvim/fileio.c:
```diff
@@ -2781,20 +2781,12 @@ buf_write (
-      int bfd;
-      char_u      *copybuf, *wp;
-      int some_error = FALSE;
+    if (backup_copy) {
```
@Said6289 I just noticed: the old code checked `os_open` result, and falls back to the `else` branch if it fails. I guess that might be worth keeping, although we have no use for the file-descriptor.

Said6289 replied:
I guess we can use a stat for that, so we don't have an open fd lying around.
